File: vuln_cmd.js
/**
 * Trigger rules:
 *   - tg-js-child-process-exec  (child_process.exec with user input)
 * CWE-78: OS Command Injection
 */
const express = require('express');
const { exec } = require('child_process');
const app = express();

// ── 🔴 exec() with string concatenation ───────────────────────────────────
app.get('/ping', (req, res) => {
  const host = req.query.host;
  exec("ping -c 3 " + host, (err, stdout) => { // ❌ RULE HIT: tg-js-child-process-exec
    res.send(stdout);
  });
});

// ── 🔴 exec() with template literal ───────────────────────────────────────
app.get('/dns', (req, res) => {
  const domain = req.query.domain;
  exec(`nslookup ${domain}`, (err, stdout) => { // ❌ RULE HIT: tg-js-child-process-exec
    res.send(stdout);
  });
});

// ── 🔴 exec() with file operations ────────────────────────────────────────
app.get('/logs', (req, res) => {
  const logfile = req.query.file;
  exec(`cat /var/log/${logfile}`, (err, stdout) => { // ❌ RULE HIT
    res.send(stdout);
  });
});

module.exports = app;

File: vuln_cmd.py
"""
Trigger rules:
  - tg-py-os-system-injection    (os.system with dynamic input)
  - tg-py-subprocess-shell-true  (subprocess with shell=True)
CWE-78: OS Command Injection
"""
import os
import subprocess

# ── 🔴 os.system() with string concatenation ───────────────────────────────
hostname = input("Enter hostname to ping: ")
os.system("ping -c 3 " + hostname)  # ❌ RULE HIT: tg-py-os-system-injection

# ── 🔴 os.system() with f-string ───────────────────────────────────────────
filename = input("Enter file to remove: ")
os.system(f"rm -rf /tmp/{filename}")  # ❌ RULE HIT: tg-py-os-system-injection

# ── 🔴 os.system() with .format() ──────────────────────────────────────────
ip_addr = input("Enter IP: ")
os.system("nslookup {}".format(ip_addr))  # ❌ RULE HIT: tg-py-os-system-injection

# ── 🔴 subprocess.call with shell=True ─────────────────────────────────────
user_cmd = input("Enter command: ")
subprocess.call(user_cmd, shell=True)  # ❌ RULE HIT: tg-py-subprocess-shell-true

# ── 🔴 subprocess.Popen with shell=True ────────────────────────────────────
log_path = input("Enter log file: ")
proc = subprocess.Popen(f"tail -f {log_path}", shell=True)  # ❌ RULE HIT

# ── 🔴 subprocess.run with shell=True ──────────────────────────────────────
domain = input("Enter domain: ")
subprocess.run(f"dig {domain}", shell=True, capture_output=True)  # ❌ RULE HIT
